File: backend/src/main/java/com/agentworkbench/harness/core/EnvironmentInfoProvider.java
package com.agentworkbench.harness.core;

import com.agentworkbench.session.entity.Session;
import lombok.Builder;
import org.springframework.stereotype.Component;

import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.Locale;
import java.util.concurrent.TimeUnit;

/**
 * Collects execution environment details for prompt context.
 */
@Component
public class EnvironmentInfoProvider {

    public EnvironmentInfo detect(String workspace) {
        return EnvironmentInfo.builder()
                .isGit(isGitWorkspace(workspace))
                .platform(normalizePlatform(System.getProperty("os.name")))
                .shell(resolveShell())
                .osVersion(buildOsVersion())
                .build();
    }

    public EnvironmentInfo fromSessionOrDetect(Session session) {
        if ("LOCAL".equalsIgnoreCase(session.getExecutionMode())) {
            return EnvironmentInfo.builder()
                    .isGit(session.getIsGit())
                    .platform(session.getPlatform())
                    .shell(session.getShellPath())
                    .osVersion(session.getOsVersion())
                    .build();
        }

        EnvironmentInfo detected = detect(session.getWorkspace());
        return EnvironmentInfo.builder()
                .isGit(session.getIsGit() != null ? session.getIsGit() : detected.isGit())
                .platform(hasText(session.getPlatform()) ? session.getPlatform() : detected.platform())
                .shell(hasText(session.getShellPath()) ? session.getShellPath() : detected.shell())
                .osVersion(hasText(session.getOsVersion()) ? session.getOsVersion() : detected.osVersion())
                .build();
    }

    private boolean isGitWorkspace(String workspace) {
        if (!hasText(workspace)) return false;
        Path current = Path.of(workspace).toAbsolutePath().normalize();
        while (current != null) {
            if (Files.exists(current.resolve(".git"))) {
                return true;
            }
            current = current.getParent();
        }
        return false;
    }

    private String normalizePlatform(String osName) {
        String normalized = osName != null ? osName.toLowerCase(Locale.ROOT) : "";
        if (normalized.contains("mac") || normalized.contains("darwin")) return "darwin";
        if (normalized.contains("win")) return "win32";
        return "linux";
    }

    private String resolveShell() {
        String comspec = System.getenv("COMSPEC");
        if (hasText(comspec)) return comspec;
        return "win32".equals(normalizePlatform(System.getProperty("os.name"))) ? "cmd.exe" : "bash";
    }

    private String buildOsVersion() {
        String platform = normalizePlatform(System.getProperty("os.name"));
        String osName = System.getProperty("os.name", "");
        String osVersion = System.getProperty("os.version", "");
        if ("darwin".equals(platform) || "linux".equals(platform)) {
            String uname = readUnameSr();
            if (hasText(uname)) return uname;
        }
        if ("darwin".equals(platform)) {
            return "Darwin " + osVersion;
        }
        if ("linux".equals(platform)) {
            return "Linux " + osVersion;
        }
        return osName + " " + osVersion;
    }

    private String readUnameSr() {
        try {
            Process process = new ProcessBuilder("uname", "-sr")
                    .redirectErrorStream(true)
                    .start();
            if (!process.waitFor(2, TimeUnit.SECONDS)) {
                process.destroyForcibly();
                return null;
            }
            try (BufferedReader reader = new BufferedReader(new InputStreamReader(process.getInputStream()))) {
                return reader.readLine();
            }
        } catch (Exception ignored) {
            return null;
        }
    }

    private boolean hasText(String value) {
        return value != null && !value.isBlank();
    }

    @Builder
    public record EnvironmentInfo(Boolean isGit, String platform, String shell, String osVersion) {}
}

File: backend-ts/src/harness/core/environment-info-provider.ts
import { execFile } from 'node:child_process';
import { existsSync } from 'node:fs';
import os from 'node:os';
import path from 'node:path';
import { promisify } from 'node:util';
import { hasText } from '../../common/case.js';
import type { Session } from '../deps.js';

const execFileAsync = promisify(execFile);

export interface EnvironmentInfo {
  isGit?: boolean | null;
  platform?: string | null;
  shell?: string | null;
  osVersion?: string | null;
}

export class EnvironmentInfoProvider {
  async detect(workspace: string | null | undefined): Promise<EnvironmentInfo> {
    return {
      isGit: this.isGitWorkspace(workspace),
      platform: this.normalizePlatform(os.platform()),
      shell: this.resolveShell(),
      osVersion: await this.buildOsVersion(),
    };
  }

  async fromSessionOrDetect(session: Session): Promise<EnvironmentInfo> {
    if (session.executionMode?.toUpperCase() === 'LOCAL') {
      return {
        isGit: session.isGit === true || session.isGit === 1,
        platform: session.platform,
        shell: session.shellPath,
        osVersion: session.osVersion,
      };
    }
    const detected = await this.detect(session.workspace);
    return {
      isGit: session.isGit != null ? (session.isGit === true || session.isGit === 1) : detected.isGit,
      platform: hasText(session.platform) ? session.platform : detected.platform,
      shell: hasText(session.shellPath) ? session.shellPath : detected.shell,
      osVersion: hasText(session.osVersion) ? session.osVersion : detected.osVersion,
    };
  }

  private isGitWorkspace(workspace: string | null | undefined): boolean {
    if (!hasText(workspace)) return false;
    let current: string | null = path.resolve(workspace!);
    while (current) {
      if (existsSync(path.join(current, '.git'))) return true;
      const parent = path.dirname(current);
      if (parent === current) break;
      current = parent;
    }
    return false;
  }

  private normalizePlatform(osName: string): string {
    const normalized = (osName ?? '').toLowerCase();
    if (normalized.includes('mac') || normalized.includes('darwin')) return 'darwin';
    if (normalized.includes('win')) return 'win32';
    return 'linux';
  }

  private resolveShell(): string {
    const comspec = process.env.COMSPEC;
    if (hasText(comspec)) return comspec!;
    return this.normalizePlatform(os.platform()) === 'win32' ? 'cmd.exe' : 'bash';
  }

  private async buildOsVersion(): Promise<string> {
    const platform = this.normalizePlatform(os.platform());
    const osName = os.type();
    const osVersion = os.release();
    if (platform === 'darwin' || platform === 'linux') {
      const uname = await this.readUnameSr();
      if (hasText(uname)) return uname!;
    }
    if (platform === 'darwin') return 'Darwin ' + osVersion;
    if (platform === 'linux') return 'Linux ' + osVersion;
    return osName + ' ' + osVersion;
  }

  private async readUnameSr(): Promise<string | null> {
    try {
      const { stdout } = await execFileAsync('uname', ['-sr'], { timeout: 2000 });
      return stdout.trim().split('\n')[0] ?? null;
    } catch {
      return null;
    }
  }
}
